File: modules/kitchen-spatial/android/src/main/java/com/taqueriaelreloj/kitchenspatial/KitchenSpatialModule.kt
package com.taqueriaelreloj.kitchenspatial

import com.google.ar.core.ArCoreApk
import com.google.ar.core.Config
import com.google.ar.core.Session
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class KitchenSpatialModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("KitchenSpatial")

    AsyncFunction("getCapabilities") {
      val context = requireNotNull(appContext.reactContext)
      val availability = ArCoreApk.getInstance().checkAvailability(context)
      var depthSupported = false
      if (availability.isSupported && availability == ArCoreApk.Availability.SUPPORTED_INSTALLED) {
        depthSupported = try {
          val session = Session(context)
          try {
            session.isDepthModeSupported(Config.DepthMode.AUTOMATIC)
          } finally {
            session.close()
          }
        } catch (_: Exception) {
          false
        }
      }
      mapOf(
        "provider" to if (depthSupported) "arcore-depth" else "guided-camera",
        "roomPlanSupported" to false,
        "arCoreSupported" to availability.isSupported,
        "depthSupported" to depthSupported,
        "status" to availability.name
      )
    }
  }
}

File: modules/kitchen-spatial/ios/KitchenSpatialModule.swift
import ExpoModulesCore
import RoomPlan

public class KitchenSpatialModule: Module {
  public func definition() -> ModuleDefinition {
    Name("KitchenSpatial")

    AsyncFunction("getCapabilities") { () -> [String: Any] in
      let supported = RoomCaptureSession.isSupported
      return [
        "provider": supported ? "roomplan" : "guided-camera",
        "roomPlanSupported": supported,
        "arCoreSupported": false,
        "depthSupported": supported,
        "status": supported ? "SUPPORTED" : "LIDAR_NOT_AVAILABLE"
      ]
    }
  }
}
